packages/zk-eddsa-event-ticket-pcd/src/ZKEdDSAEventTicketPCD.ts

```diff
@@ -0,0 +1,538 @@
+import { EDdSAPublicKey } from "@pcd/eddsa-pcd";
+import {
+  EdDSATicketPCD,
+  EdDSATicketPCDPackage,
+  ITicketData,
+  ticketDataToBigInts
+} from "@pcd/eddsa-ticket-pcd";
+import {
+  BigIntArgument,
+  DisplayOptions,
+  ObjectArgument,
+  PCD,
+  PCDArgument,
+  PCDPackage,
+  SerializedPCD,
+  StringArrayArgument
+} from "@pcd/pcd-types";
+import {
+  SemaphoreIdentityPCD,
+  SemaphoreIdentityPCDPackage
+} from "@pcd/semaphore-identity-pcd";
+import { STATIC_SIGNATURE_PCD_NULLIFIER } from "@pcd/semaphore-signature-pcd";
+import {
+  BABY_JUB_NEGATIVE_ONE,
+  babyJubIsNegativeOne,
+  booleanToBigInt,
+  decStringToBigIntToUuid,
+  fromHexString,
+  generateSnarkMessageHash,
+  uuidToBigInt
+} from "@pcd/util";
+// eslint-disable-next-line @typescript-eslint/triple-slash-reference
+/// <reference path="../../util/src/declarations/circomlibjs.d.ts" />
+// eslint-disable-next-line @typescript-eslint/triple-slash-reference
+/// <reference path="../../util/src/declarations/snarkjs.d.ts" />
+import { BabyJub, buildBabyjub, buildEddsa, Eddsa } from "circomlibjs";
+import JSONBig from "json-bigint";
+import { groth16 } from "snarkjs";
+import { v4 as uuid } from "uuid";
+import vkey from "../artifacts-unsafe/verification_key.json";
+
+import { ZKEdDSAEventTicketCardBody } from "./CardBody";
+
+export const STATIC_TICKET_PCD_NULLIFIER = generateSnarkMessageHash(
+  "dummy-nullifier-for-eddsa-event-ticket-pcds"
+);
+
+export const ZKEdDSAEventTicketPCDTypeName = "zk-eddsa-event-ticket-pcd";
+
+let initializedPromise: Promise<void> | undefined;
+let babyJub: BabyJub;
+let eddsa: Eddsa;
+let initArgs: ZKEdDSAEventTicketPCDInitArgs | undefined = undefined;
+
+/**
+ * Specifies which fields of an EdDSATicket should be revealed in a proof.
+ */
+export interface EdDSATicketFieldsToReveal {
+  revealTicketId?: boolean;
+  revealEventId?: boolean;
+  revealProductId?: boolean;
+  revealTimestampConsumed?: boolean;
+  revealTimestampSigned?: boolean;
+  revealAttendeeSemaphoreId?: boolean;
+  revealIsConsumed?: boolean;
+  revealIsRevoked?: boolean;
+}
+
+/**
+ * Info required to initialize this PCD package.  These are the artifacts
+ * associated with the circom circuit.
+ */
+export interface ZKEdDSAEventTicketPCDInitArgs {
+  zkeyFilePath: string;
+  wasmFilePath: string;
+}
+
+/**
+ * Max supported size of validEventIds field in ZKEdDSAEventTicketPCDArgs.
+ */
+export const VALID_EVENT_IDS_MAX_LEN = 100;
+
+/**
+ * Arguments to request a new proof.
+ */
+export interface ZKEdDSAEventTicketPCDArgs {
+  // generally, `ticket` and `identity` are user-provided
+  ticket: PCDArgument<EdDSATicketPCD>;
+  identity: PCDArgument<SemaphoreIdentityPCD>;
+
+  // `validEventIds` is usually app-specified.  It is optional, and if included
+  // the PCD proves that the ticket's event ID is in this list.  This is a list of
+  // UUIDs with max length VALID_EVENT_IDS_MAX_LEN (100).
+  validEventIds?: StringArrayArgument;
```

Comment:
args can have optional values, but are themselves not optional. not sure how to have communicated that appropriately.
```suggestion
  validEventIds: StringArrayArgument;
```

Reply:
@ichub currently i have an optional arg `externalNullifier` in `ZKEdDSATicketPCDArgs`, and in my `prove` function if the arg is not present i default to some default value. should i not be doing that in args as well?